In [1]:
import pandas as pd
df = pd.read_csv("messy_sales.csv")
display(df.shape)
display(df.head(10))
display(df.dtypes)

(300, 6)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


order_id      int64
date         object
product      object
price       float64
qty           int64
zip           int64
dtype: object

## 3 Problems I see with the above data:

- data type for "date" is reading as an object
- data type for "zip" is reading as an integer
- there are values of NaN for some prices

In [2]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [3]:
fill_value = df["price"].median()
df["price"] = df["price"].fillna(fill_value)
print(fill_value)

37.53


In [4]:
df.isna().sum()

order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [5]:
print(df.duplicated().sum())
df = df.drop_duplicates()
df.shape

8


(292, 6)

In [6]:
df["zip"] = df["zip"].astype(str).str.zfill(5)

In [7]:
df["date"] = pd.to_datetime(df["date"], format="mixed")
display(df.dtypes)

order_id             int64
date        datetime64[ns]
product             object
price              float64
qty                  int64
zip                 object
dtype: object

In [8]:
display(df.shape)
display(df.head(10))
display(df.dtypes)

(292, 6)

,order_id,date,product,price,qty,zip
0,1254,2026-04-06,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,2026-05-06,mug,5.50,1,10001
3,1066,2026-05-30,notebook,37.53,1,98101
4,1114,2026-04-06,webcam,45.59,4,90405
5,1280,2026-04-22,pen set,37.53,4,30303
6,1204,2026-04-18,charger,8.41,2,02134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


order_id             int64
date        datetime64[ns]
product             object
price              float64
qty                  int64
zip                 object
dtype: object

In [9]:
import numpy as np
df[df["qty"] < 0]
df["neg_qty"] = np.where(df["qty"] < 0, 1, 0)
df[df["qty"] < 0]


,order_id,date,product,price,qty,zip,neg_qty
202,1140,2026-04-18,webcam,38.69,-5,98101,1
262,1233,2026-05-09,desk lamp,22.64,-4,10001,1
297,1025,2026-04-27,keyboard,47.30,-2,02116,1


In [10]:
df.to_csv('sales_clean.csv', index = False)

## Cleaning Log

300 loaded; 12 prices filled with _37.53_; 8 duplicates removed leaving 292; zips restored; dates standardized; negatives handled by flagging the rows that have a negative qty. The reason for this is that if it is showing a negative qty, then it needs to be investigated as to why a negative qty is showing. It it is just deleted, the error can continue to persist into the future. 

https://github.com/skeown13/cs82a-data-science-portfolio/tree/main